app/models/event.rb

```diff
@@ -20,18 +20,52 @@ class Event < ActiveRecord::Base
 
   accepts_nested_attributes_for :users
   accepts_nested_attributes_for :user_events
-  validates_presence_of :subject
+  validates_presence_of :subject, :datetime, :location
+  validate :datetime_no_less_than_15_mins_before_now
 
-  after_create :send_email
-  after_update :send_email
+  after_create :send_default_email
+  after_update :update_db, :send_updated_email
 
-  def send_email
+  def send_email(run_at_time, email_type)
     event = Event.find(id)
-    event_time = (event.datetime.to_time - 3.hours).to_datetime
-    users = User.all
-    users.each do |u|
-      UserMailer.delay(run_at: event_time)
-                .notification_email(u, event)
+    if event.try(:confirmed?)
+      users = User.all
+      users.each do |user|
+        UserMailer.delay(run_at: run_at_time)
+                  .send(email_type, user, event)
+      end
+    end
+  end
+
+  def send_default_email
+    if datetime - 2.hours <= 3.hours.from_now + 2.hours
+      send_email(DateTime.now, 'instant_email')
+    else
+      send_email((datetime.to_time - 3.hours).to_datetime,
+                 'notification_email')
+    end
+  end
+
+  def update_db
+    jobs = Delayed::Job.all
+    jobs.each do |job|
+      job.delete if job.handler.include? id.to_s
+    end
+  end
+
+  def send_updated_email
+    if datetime - 2.hours <= 3.hours.from_now + 2.hours
+      send_email(DateTime.now, 'instant_email')
+    else
+      send_email(DateTime.now, 'info_email')
+      send_email((datetime.to_time - 3.hours).to_datetime,
+                 'notification_email')
+    end
+  end
+
+  def datetime_no_less_than_15_mins_before_now
+    if !datetime.nil? && datetime <= 15.minutes.from_now - 2.hours
```

Comment:
1 Rubocop: 'Avoid the use of double negation (!!)'.
2
```
2.2.3 :007 > DateTime.now
 => Mon, 22 Feb 2016 11:48:52 +0200 
2.2.3 :008 > 15.minutes.from_now
 => Mon, 22 Feb 2016 10:03:53 UTC +00:00 
2.2.3 :009 > 15.minutes.from_now + 2.hours
 => Mon, 22 Feb 2016 12:03:57 UTC +00:00 
```
Thanks to time zones (?) I needed to check it like that. Dunno how to avoid.